plugins/html-plugin/src/main/java/ru/sbtqa/tag/pagefactory/find/HtmlFindUtils.java

```diff
@@ -0,0 +1,382 @@
+     *
+     * @param <T> expected type of the returned element
+     * @param name element name or path to it
+     * @param clazz list of valid types. If the element found does not match the
+     * valid type, an exception will be thrown
+     * @param wait to wait for the element to appear or not
+     * @return Returns an element from a page by name or path
+     */
+    public <T extends WebElement> T find(String name, List<Class> clazz, boolean wait) {
+        T element = find(name, wait);
+        if (!clazz.contains(element.getClass())) {
+            throw new AutotestError(format("Uncorrected element type. Element: %s. Valid types: %s", name, clazz.toString()));
+        }
+        return element;
+    }
+
+    private void findElement(Field field, ComplexElement element) throws ElementDescriptionException {
+        Object fieldObject = getElementByField(element, field);
+
+        if (field.getType().isAssignableFrom(List.class)) {
+            findElementOfList((List<WebElement>) fieldObject, element);
+        } else {
+            element.setElement((WebElement) fieldObject);
+        }
+    }
+
+    private Field getField(Object element, String elementName) {
+        Page page = PageContext.getCurrentPage();
+        Class clazz = element == null ? page.getClass() : element.getClass();
+
+        for (Field field : FieldUtilsExt.getFieldsListWithAnnotation(clazz, ElementTitle.class)) {
+            field.setAccessible(true);
+            if (field.getAnnotation(ElementTitle.class).value().equals(elementName)) {
+                return field;
+            }
+        }
+        return null;
+    }
+
+    private static void findElementOfList(List<WebElement> list, ComplexElement currentElement) {
+        int position = currentElement.getCurrentPosition();
+
+        int index = 0;
+        List<String> elementPath = currentElement.getElementPath();
+
+        if (position + 1 < elementPath.size()) {
+            String nextParam = elementPath.get(position + 1);
+            boolean isNumeric = nextParam.chars().allMatch(Character::isDigit);
+            if (isNumeric) {
+                index = Integer.parseInt(elementPath.get(position + 1)) - 1;
+                currentElement.setCurrentPosition(position + 1);
+            } else {
+                LOG.info("The parameter following the list is not an index. Index list element: 0.");
+            }
+        }
+        if (index >= list.size()) {
+            boolean isLastElement = currentElement.getCurrentPosition() + 1 != currentElement.getElementPath().size();
+            if (!isLastElement && currentElement.isWaitAppear()) {
+                throw new AutotestError(format("Index element in list "
+                        + "exceeds list size. Index: %s. Size: %s", index, list.size()));
+            } else {
+                currentElement.setElement(null);
+            }
+        } else {
+            LOG.debug("Element with index " + index + " was received.");
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Format specifiers should be used instead of string concatenation. ([squid:S3457](https://sonarcloud.io/organizations/sbtqasonarkey/rules#rule_key=squid%3AS3457))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=ru.sbtqa.tag%3Apage-factory-2&pullRequest=133&issues=AWfPt2o64tUZZzFFj6Hb&open=AWfPt2o64tUZZzFFj6Hb)

[ISSUE_KEY]: AWfPt2o64tUZZzFFj6Hb